smolyay/samples.py

```diff
@@ -0,0 +1,415 @@
+import abc
+import numpy
+
+
+class UnidimensionalPointSet(abc.ABC):
+    """Set of unidimensional points
+
+    A set of unique unidimensional points within a domain.
+    """
+
+    def __init__(self, domain):
+        self._points = None
+        self._domain = None
+
+        self.domain = numpy.array(domain, dtype=float)
+        self._valid_cache = False
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `points`."""
+        return self._domain
+
+    @domain.setter
+    def domain(self, value):
+        domain = numpy.array(value, dtype=float)
+        if domain.ndim != 1 or len(domain) != 2:
+            raise TypeError("Domain must be array with two variables")
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            self._valid_cache = False
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points in the set."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._points
+
+    def __len__(self):
+        return len(self.points)
+
+    @abc.abstractmethod
+    def _create(self):
+        """Create the points in the set."""
+        pass
+
+    def _scale_to_domain(self, points, old_domain):
+        points = self.domain[0] + (self.domain[1] - self.domain[0]) * (
+            (points - old_domain[0]) / (old_domain[1] - old_domain[0])
+        )
+        numpy.clip(points, self.domain[0], self.domain[1], out=points)
+        return points
+
+
+class NestedUnidimensionalPointSet(UnidimensionalPointSet):
+    """Set of unidimensional points assigned to levels.
+
+
+    Parameters
+    ----------
+    max_level ; int
+        the maximum level the points are used for
+    """
+
+    def __init__(self, domain, max_level):
+        super().__init__(domain)
+        self._max_level = None
+        self._num_points_per_level = None
+        self._start_level = None
+        self._end_level = None
+
+        self.max_level = max_level
+
+    @property
+    def max_level(self):
+        """int: maximum level to compute points for."""
+        return self._max_level
+
+    @max_level.setter
+    def max_level(self, value):
+        max_level = int(value)
+        if max_level != self._max_level:
+            self._max_level = max_level
+            self._valid_cache = False
+
+    @property
+    def num_points_per_level(self):
```

Comment:
I think you had suggested this in a comment before? I liked this idea to shorten the name
```suggestion
    def num_per_level(self):
```